<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
  <style>
    li {margin: .3em 0}
  </style>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>SNP calling pipeline (Mpileup/Freebayes)</title>
    	<link rel="stylesheet" type="text/css" href="mpsfor.css">
	   <script src="mynav.js" type="text/javascript"> </script>

  </head>
  <body>
<script type="text/javascript">myheader();</script>

<h2>SNP calling pipeline (Mpileup/Freebayes)</h2>

<p>SNP calling can be performed within the MPS Forensics application by 
clicking on the "SNP calling" tab, next to the "Home" tab. This should 
show you a set of options and sub-tabs like this:
</p>
<p><img src="images/snp_pipeline.png"></img></p>
<p>From here it is possible to choose the data set you wish to use. By 
default the pipelines will search for matching paired-end data and use 
this if found, you can override this behaviour and force single or 
paired-end operation. Reads are first QC'd and low quality bases/reads 
removed with <a href="http://www.usadellab.org/cms/?page=trimmomatic">
Trimmomatic</a>, before a second round of QC results are generated.</p>

<p>All standard tools for SNP finding use the same 
method of aligning reads to a reference genome - hg19 is the default, 
but another, indexed by BWA may be chosen as a "custom reference" - 
and then calling the differences that have been found using a dedicated 
tool. The pipeline installed here uses the 
<a href="https://arxiv.org/abs/1303.3997">BWA mem algorithm</a> to 
perform the alignment step. Users can chose from 
<a href="http://samtools.sourceforge.net/mpileup.shtml">Samtools 
Mpileup</a> or <a href="https://github.com/ekg/freebayes">Freebayes</a> 
for calling SNPs.</p>

<p>Note: depending on the options chosen for each of these tools, it 
is possible to call indels in addition to SNPs.</p>

<script type="text/javascript" > myfooter(); </script>
  </body>
</html>
