<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>STRait Razor v2.5 pipeline</title>
</head>
<body>

<a href="help.html"><img src="images/back.png">&nbsp;&nbsp;
Return to help page</img></a>

<h2>STRait Razor v2.5 pipeline</h2>
<p>The STRait Razor v2.5 (from here-on referred to in short as Razor) 
pipeline identifies STRs from DNA sequences by running an approximate 
text matching algorithm (tre-agrep) on each sequence. It is capable 
of working with paired-end read data, though the reads must be long 
enough to sequence through the entire repeat region.</p>
<p>STRait Razor is capable of identifying STRs from almost every 
legimate STR-amplicon read, but it has no allelotype (genotype) caller 
available with it, which makes the output from Razor hard to transition 
into our allele viewer and no genotype is called, so we cannot further 
summarise the outputs.</p>
<img src="images/straitrazor_pipeline.png"></img>
<p>STRait Razor is poorly written and offers no control of the number 
of threads it uses - always maxing out the host CPU. We have also had 
to modify its code to:
<ol><li>handle gzipped fastq files</li>
<li>write to locations other than its own</li>
<li>operate from a different location to its data</li></ol>
These changes are located in the /mpsforensics/forensicsapp/STRaitRazor.patch 
file. Upon installation the patch is applied to the user's installation 
of STRait Razor v2.5.</p>

<a href="help.html"><img src="images/back.png">&nbsp;&nbsp;
Return to help page</img></a>

</body>
</html>
